cfme/fixtures/authentication.py

```diff
@@ -0,0 +1,97 @@
+    appliance.server.authentication.configure(auth_mode='amazon',
+                                              auth_provider=amazon_auth_provider)
+    yield
+
+    appliance.server.authentication.auth_settings = original_config
+    appliance.server.login_admin()
+    appliance.server.authentication.configure(auth_mode='database')
+
+
+@pytest.fixture(scope='function')
+def auth_provider(prov_key):
+        return authutil.get_auth_crud(prov_key)
+
+
+@pytest.fixture(scope='function')
+def auth_user_data(auth_provider, user_type):
+    """Grab user data attrdict from auth provider's user data in yaml
+
+    Expected formatting of yaml containing user data:
+        test_users:
+        -
+          username: ldapuser2
+          password: mysecretpassworddontguess
+          fullname: Ldap User2
+          groupname: customgroup1
+          providers:
+            - freeipa01
+          user_types:
+            - uid
+
+    Only include user data for users where the user_type matches that under test
+
+    Assert the data isn't empty, and skip the test if so
+    """
+    try:
+        data = [user
+                for user in auth_provider.user_data
+                if user_type in user.user_types]
+        assert data
+    except (KeyError, AttributeError, AssertionError):
+        logger.exception('Exception fetching auth_user_data from yaml')
+        pytest.skip('No yaml data for auth_prov {} under "auth_data.test_data"'
+                    .format(auth_provider))
+    return data
+
+
+@pytest.yield_fixture(scope='function')
+def configure_auth(appliance, auth_mode, auth_provider, user_type, request):
+    """Given auth_mode, auth_provider, user_type parametrization, configure auth for login
+    testing.
+
+    Saves original auth settings
+    Configures external or internal auth modes
+    Separate freeipa / openldap config methods and finalizers
+    Restores original auth settings after yielding
+    """
+    original_config = appliance.server.authentication.auth_settings
+    logger.debug('Original auth settings before configure_auth fixture: %r', original_config)
+    if auth_mode.lower() != 'external':
+        appliance.server.authentication.configure(auth_mode=auth_mode,
+                                                  auth_provider=auth_provider,
+                                                  user_type=user_type)
+    elif auth_mode.lower() == 'external':  # extra explicit
+        if auth_provider.auth_type == 'freeipa':
+            appliance.configure_freeipa(auth_provider)
+            request.addfinalizer(appliance.disable_freeipa)
+        elif auth_provider.auth_type == 'openldaps':
+            appliance.configure_openldap(auth_provider)
+            request.addfinalizer(appliance.disable_openldap)
+
+    # Auth reconfigure is super buggy and sensitive to timing
+    # Just waiting on sssd to be running, or an httpd restart isn't sufficient
+    sleep(30)
```

Comment:
**REQUIRED:** So can we instead wait for SSSD to run?

Reply:
Comment above mentions, I tried waiting for both httpd and sssd, and still saw login failures after the services were up. I'm not sure what else to wait on in this context, but a 30s pause before continuing and trying to login with the newly configured auth made things consistent run-over-run.